src/Transaction.cc

```diff
@@ -0,0 +1,293 @@
+ * \param buf
+ *      Address of the first byte of the new contents for the object;
+ *      must contain at least length bytes.
+ * \param length
+ *      Size in bytes of the new contents for the object.
+ */
+void
+Transaction::write(uint64_t tableId, const void* key, uint16_t keyLength,
+        const void* buf, uint32_t length)
+{
+    if (expect_false(commitStarted)) {
+        throw TxOpAfterCommit(HERE);
+    }
+
+    ClientTransactionTask* task = taskPtr.get();
+
+    Key keyObj(tableId, key, keyLength);
+    ClientTransactionTask::CacheEntry* entry = task->findCacheEntry(keyObj);
+
+    if (entry == NULL) {
+        entry = task->insertCacheEntry(tableId, key, keyLength, buf, length);
+    } else {
+        entry->objectBuf->reset();
+        Object::appendKeysAndValueToBuffer(
+                keyObj, buf, length, entry->objectBuf, true);
+    }
+
+    entry->type = ClientTransactionTask::CacheEntry::WRITE;
+}
+
+/**
+ * Constructor for Transaction::ReadOp: initiates a read just like
+ * #Transaction::read, but returns once the operation has been initiated,
+ * without waiting for it to complete.  The operation is not consider part of
+ * the transaction until it is waited on.
+ *
+ * \param transaction
+ *      The Transaction object of which this operation is a part.
+ * \param tableId
+ *      The table containing the desired object (return value from
+ *      a previous call to getTableId).
+ * \param key
+ *      Variable length key that uniquely identifies the object within tableId.
+ *      It does not necessarily have to be null terminated.
+ * \param keyLength
+ *      Size in bytes of the key.
+ * \param[out] value
+ *      After a successful return, this Buffer will hold the
+ *      contents of the desired object - only the value portion of the object.
+ */
+Transaction::ReadOp::ReadOp(Transaction* transaction, uint64_t tableId,
+        const void* key, uint16_t keyLength, Buffer* value)
+    : transaction(transaction)
+    , tableId(tableId)
+    , keyBuf()
+    , keyLength(keyLength)
+    , value(value)
+    , buf()
+    , rpc()
+{
+    keyBuf.appendCopy(key, keyLength);
+
+    ClientTransactionTask* task = transaction->taskPtr.get();
+
+    Key keyObj(tableId, key, keyLength);
+    ClientTransactionTask::CacheEntry* entry = task->findCacheEntry(keyObj);
+
+    // If no cache entry exists an rpc should be issued.
+    if (entry == NULL) {
+        rpc.construct(
+                transaction->ramcloud, tableId, key, keyLength, &buf);
+    }
+    // Otherwise we will just return it from cache when wait is called.
+}
+
+/**
+ * Wait for the operation to complete.  The operation is not part of the
+ * transaction until wait is called (e.g. if commit is called before wait,
+ * this operation will not be included).  Behavior when calling wait more than
+ * once is undefined.
+ */
+void
+Transaction::ReadOp::wait()
+{
+    if (expect_false(transaction->commitStarted)) {
+        throw TxOpAfterCommit(HERE);
+    }
```

Comment:
Shouldn't you throw this exception in the constructor, rather than here? E.g. iis it a good idea to start a read RPC after the transaction commit has started?